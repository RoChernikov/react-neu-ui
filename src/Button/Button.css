.button {
  color: #fff;
  padding: 0.5rem 2rem;
  border: none;
  border-radius: 0.25rem;
}

.button_primary {
  background-color: rgb(0, 132, 255);
}

.button_success {
  background-color: rgb(0, 255, 115);
}
