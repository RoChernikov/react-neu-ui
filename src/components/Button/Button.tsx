import React, { FC, ReactNode } from 'react';
import { clsx } from 'clsx';
import styles from './Button.module.css';

export type TButtonProps = {
  onClick(): void;
  children: ReactNode;
  variant?: 'primary' | 'success' | 'warning';
  isDisabled?: boolean;
};

const Button: FC<TButtonProps> = ({
  onClick,
  children,
  variant = 'primary',
  isDisabled = false,
}) => {
  return (
    <button
      className={clsx({
        [styles.button]: true,
        [styles.button_primary]: variant === 'primary',
        [styles.button_success]: variant === 'success',
        [styles.button_warning]: variant === 'warning',
      })}
      onClick={onClick}
      disabled={isDisabled}
    >
      {children}
    </button>
  );
};

export default Button;
