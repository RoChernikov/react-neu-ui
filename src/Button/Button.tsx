import React, { FC, ReactNode } from 'react';
import './Button.css';

export type TButtonProps = {
  onClick(): void;
  children: ReactNode;
  variant?: 'primary' | 'success';
  isDisabled?: boolean;
};

const Button: FC<TButtonProps> = ({
  onClick,
  children,
  variant = 'primary',
  isDisabled = false,
}) => {
  const cn = `button button_${variant}`;
  return (
    <button className={cn} onClick={onClick} disabled={isDisabled}>
      {children}
    </button>
  );
};

export default Button;
