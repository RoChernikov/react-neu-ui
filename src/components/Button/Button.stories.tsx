import React from 'react';
import { Meta, Story } from '@storybook/react';
import Button, { TButtonProps } from './Button';

const meta: Meta = {
  title: 'Welcome',
  component: Button,
  argTypes: {
    children: {
      control: {
        type: 'text',
      },
    },
  },
  parameters: {
    controls: { expanded: true },
  },
};

export default meta;

const Template: Story<TButtonProps> = args => <Button {...args} />;

// By passing using the Args format for exported stories, you can control the props for a component for reuse in a test
// https://storybook.js.org/docs/react/workflows/unit-testing
export const RegularButton = Template.bind({});

RegularButton.args = {
  onClick: () => {
    alert('CLICKED!');
  },
  children: 'Primary button',
  variant: 'primary',
};
