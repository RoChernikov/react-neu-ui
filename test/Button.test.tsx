import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import Button from '../src/Button/Button';

describe('Button', () => {
  describe('should be rendered correctly', () => {
    it('default', () => {
      render(<Button onClick={console.log}>Click me</Button>);
      const button = screen.getByText('Click me');
      expect(button.classList.contains('button_primary')).toBeTruthy();
    });
    it('success', () => {
      const { container } = render(
        <Button onClick={console.log} variant="success">
          Click me
        </Button>
      );
      const button = container.querySelector('button');
      expect(button!.classList.contains('button_success')).toBeTruthy();
    });
    it('disabled', () => {
      const { asFragment } = render(
        <Button onClick={console.log} isDisabled={true}>
          Click me
        </Button>
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('should be clickable', () => {
    const onClick = jest.fn();
    render(<Button onClick={onClick}>Click me</Button>);
    fireEvent.click(screen.getByText('Click me'));
    expect(onClick).toHaveBeenCalledTimes(1);
  });
  it('should not be clickable if disabled', () => {
    const onClick = jest.fn();
    render(
      <Button onClick={onClick} isDisabled={true}>
        Click me
      </Button>
    );
    fireEvent.click(screen.getByText('Click me'));
    expect(onClick).not.toHaveBeenCalled();
  });
});
