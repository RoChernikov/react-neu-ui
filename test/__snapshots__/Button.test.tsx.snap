// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should be rendered correctly disabled 1`] = `
<DocumentFragment>
  <button
    class="button button_primary"
    disabled=""
  >
    Click me
  </button>
</DocumentFragment>
`;
